crates/accelerate/src/commutation_analysis.rs

```diff
@@ -0,0 +1,197 @@
+                            packed_inst0.extra_attrs.as_deref(),
+                            qargs1,
+                            cargs1,
+                            &op2,
+                            params2,
+                            packed_inst1.extra_attrs.as_deref(),
+                            qargs2,
+                            cargs2,
+                            max_num_qubits,
+                        )?;
+                        if !all_commute {
+                            break;
+                        }
+                    } else {
+                        all_commute = false;
+                        break;
+                    }
+                }
+
+                if all_commute {
+                    // all commute, add to current list
+                    last.insert(current_gate_idx);
+                } else {
+                    // does not commute, create new list
+                    commutation_entry.push(AIndexSet::from_iter([current_gate_idx]))
+                }
+            }
+
+            node_indices.insert(
+                (current_gate_idx, wire.clone()),
+                commutation_entry.len() - 1,
+            );
+        }
+    }
+
+    Ok((commutation_set, node_indices))
+}
+
+#[pyfunction]
+#[pyo3(signature = (dag, commutation_checker))]
+pub(crate) fn analyze_commutations(
+    py: Python,
+    dag: &mut DAGCircuit,
+    commutation_checker: &mut CommutationChecker,
+) -> PyResult<Py<PyDict>> {
+    // This returns two HashMaps:
+    //   * The commuting nodes per wire: {wire: [commuting_nodes_1, commuting_nodes_2, ...]}
+    //   * The index in which commutation set a given node is located on a wire: {(node, wire): index}
+    // The Python dict will store both of these dictionaries in one.
+    let (commutation_set, node_indices) = analyze_commutations_inner(py, dag, commutation_checker)?;
+
+    let out_dict = PyDict::new_bound(py);
+
+    // First set the {wire: [commuting_nodes_1, ...]} bit
+    for (wire, commutations) in commutation_set {
+        let py_wire = match wire {
+            Wire::Qubit(q) => dag.qubits.get(q).unwrap().to_object(py),
+            Wire::Clbit(c) => dag.clbits.get(c).unwrap().to_object(py),
+            Wire::Var(v) => v,
+        };
+
+        out_dict.set_item(
+            py_wire,
+            PyList::new_bound(
+                py,
+                commutations.iter().map(|inner| {
+                    PyList::new_bound(
+                        py,
+                        inner
+                            .into_iter()
+                            .map(|node_index| dag.get_node(py, *node_index).unwrap()),
+                    )
+                }),
+            ),
+        )?;
+    }
+
+    // Then we add the {(node, wire): index} dictionary
+    for ((node_index, wire), index) in node_indices {
+        // we could cache the py_wires to avoid this match and the python object creation,
+        // but this didn't make a noticable difference in runtime
```

Comment:
Realistically the `BitData` you're accessing here is the cache. It's basically just doing a vec lookup and returning the python object.

That being said I'm wondering can wire here be anything besides a `Qubit`? Not that it really matters because this code is correct and will handle the other cases as expected.

Reply:
No you're right, it cannot, since we explicitly only iterate over `0..dag.num_qubits().map(|i| Wire::Qubit(i))`. We can change the code to expect a Qubit, that might be a bit cleaner to the readers

Reply:
I was wondering what happens if a dag has non-continuous qubits, i.e. gaps in the set of qubits such as qubits 0...3 and 6..10 with no qubits at index 4, 5 - can this happen and was this supported in the Python implementation?